### ISSUE TITLE
[BUG] regression with Intel bluetooth bt-offload tplgs

### ISSUE THREAD

**Issue by kv2019i** [thesofproject/sof-test#1185]
**Describe the bug**
Commit c77a4feb2f1f15a3a29edf49b091b4b4a507be48 breaks BT offload tplgs like sof-nocodec-bt-mtl-lbm.tplg .

**To Reproduce**
TPLG=/lib/firmware/intel/development/sof-nocodec-bt-mtl-lbm.tplg MODEL=MTLP_RVP_NOCODEC-bluetooth-offload SOF_TEST_INTERVAL=5 ~/sof-test/test-case/check-playback.sh -d 3 -l 1 -r 1 -F 

**Reproduction Rate**
100%

**Expected behavior**
Playback works.

**Impact**
Playback fails.

**Environment**
1) SOF Commit: 
[c77a4feb2f1f](https://github.com/thesofproject/sof/commit/c77a4feb2f1f15a3a29edf49b091b4b4a507be48)
2) Linux Commit: 
[4ede2a0ed114](https://github.com/thesofproject/linux/commit/4ede2a0ed114a633d8eb9868ad6ad91cf0b08f79)

2) Name of the topology file
    * Topology: sof-nocodec-bt-lbm.tplg
3) Name of the platform(s) on which the bug is observed.
    * Platform: Intel MTL

**Screenshots or console output**

```
2024-04-17 11:13:38 UTC [REMOTE_COMMAND] sof-tplgreader.py /usr/lib/firmware/intel/development/sof-nocodec-bt-mtl-lbm.tplg -f 'type:playback & id:any & ~pcm:HDMI' -b ' pcm:HDA Digital' -s 0 -e
2024-04-17 11:13:38 UTC [REMOTE_INFO] ===== Testing: (Round: 1/1) (PCM: Port2 [hw:0,2]) (Loop: 1/1) =====
2024-04-17 11:13:38 UTC [REMOTE_COMMAND] aplay   -Dhw:0,2 -r 48000 -c 1 -f S16_LE -d 3 /dev/zero -v -q
aplay: set_params:1416: Unable to install hw params:
ACCESS:  RW_INTERLEAVED
FORMAT:  S16_LE
SUBFORMAT:  STD
SAMPLE_BITS: 16
FRAME_BITS: 16
CHANNELS: 1
RATE: 48000
PERIOD_TIME: 125000
PERIOD_SIZE: 6000
PERIOD_BYTES: 12000
PERIODS: 4
BUFFER_TIME: 500000
BUFFER_SIZE: 24000
BUFFER_BYTES: 48000
TICK_TIME: 0 
```

**Comment by marc-hb** [comment]
> (in sof-test) we are testing Bluetooth offload, and ignore known bad audio format combinations.

OK but how? In pseudo-code.

**Comment by ujfalusi** [comment]
@kv2019i, in the kernel we can have rules (with `snd_pcm_hw_rule_add()`). With that you can say that for example mono is only allowed with 16KHz (not sure what the constraint is).
Discovering or to be precise describing the constraint is another question, probably we could use the copier formats to scan through and 'rule out' combinations that is not listed.

**Comment by kv2019i** [comment]
@marc-hb I have no easy answer. Something along the lines of:
``
  if (Intel bluetooth offload)
        if (rate == 48000)
            channels = 2;
       else
            channels = 1;
```

Not sure how to implement the if condition check. This is a property of the tplg in the end. This is a clunky solution perhaps, but if the idea was to not write a completely new test for BT offload, we need to somehow figure out a way to decorate the generic test code so it can work with this specific type of tplg.

**Comment by marc-hb** [comment]
> The alsa-lib C interface allows applications to discover such complex constraints. I.e. application can leave channel count as "open" or use the "_near" variants to ask a specific value but be open to adjust to hardware capabilities. sof-test does not use this interface, but rather it implements its own discover of capability and aplay/arecord are instructed to use specific settings via the "hw:x" layer and explicit setting of all key audio parameters. This makes sense as we want to specifically test the audio drivers and want to control the audio parameters used in the tests.

It makes sense for sof-test to explicitly set parameters but it does not make sense for sof-test to be ignorant and not use the same discovery APIs as other ALSA applications. sof-test is really an ALSA test suite in disguise, so ALSA discovery APIs should be part of the sof-test coverage - and this would also provide the data that has been missing here.

In any case this looks like a major sof-test change.

**Comment by ranj063** [comment]
> 4\. Also from @ranj063, the "overkill"  solution (her name): also change Bluetooth topologies but this time add more, "narrower" and simpler pipelines (which gives more PCMs).

The more I think about it. this seems like the most logical solution. Advertising a combination of rates/channels that cannot possibly work doesn't seem like the right thing to do

**Comment by kv2019i** [comment]
@ranj063 wrote:
>> 4. Also from @ranj063, the "overkill"  solution (her name): also change Bluetooth topologies but this time add more, "narrower" and simpler pipelines (which gives more PCMs).
> 
> The more I think about it. this seems like the most logical solution. Advertising a combination of rates/channels that cannot possibly work doesn't seem like the right thing to do

But this is not according to the design we have with CRAS. We identify a single PCM for offload and that single PCM is expected to support Bluetooth audio configurations. The number of configurations will increase later, so it will not scale to have distinct PCM for all possible combinations.

The valid PCM values are standardized in the Bluetooth standards, so I'm not sure why we need to solve the discovery problem in a more generic fashion than is required for product use.

**Comment by ujfalusi** [comment]
I tend to agree with @kv2019i, here the test should know what it is testing. BT offload is a special case, needing special treatment.
No matter how much constraining, heuristics we put in kernel and/or in tplg it is not going to stop user space for asking for exactly a combination which is not supported, right?

There is always going to be a corner case and whack a mole is not fun in a long run.

But, if we really want to do this then we need topology change and kernel change to bring up the rules to aid discovery (which application can just not use and got hard for the non supported combination - again).
Coding this in tplg and using the tplg itself in CI to magically handle this? Adding a mini SOF-UCM in tplg? We cannot expect applications to read topologies. Does it really worth the effort?
